src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    internal sealed class CurlResponseMessage : HttpResponseMessage
+    {
+        private volatile bool _disposed = false;
+        private readonly CurlResponseStream _responseStream = new CurlResponseStream();
+        private readonly ManualResetEvent _headersReceivedEvent = new ManualResetEvent(false);
+
+        #region Properties
+
+        internal ReaderWriterStream ContentStream
+        {
+            get
+            {
+                return _responseStream.InnerStream;
+            }
+        }
+
+        #endregion
+
+        internal CurlResponseMessage(HttpRequestMessage request)
+        {
+            RequestMessage = request;
+            Content = new StreamContent(_responseStream);
+        }
+
+        internal void WaitForHeaders(CancellationToken cancellationToken)
+        {
+            CheckDisposed();
+            var events = new WaitHandle[] {_headersReceivedEvent, cancellationToken.WaitHandle};
+            WaitHandle.WaitAny(events, Timeout.InfiniteTimeSpan);
```

Comment:
If we end up needing this to be synchronous (which I hope we don't), then with _headersReceivedEvent as a ManualResetEventSlim instead of a ManualResetEvent, this can just be ```_headersReceivedEvent.Wait(cancellationToken)``` which is both simpler and more efficient.